Added missing includes: <netinet/in.h> and <sys/socket.h>

yurivict opened the pull request:
Found these missing includes while compiling on FreeBSD 11.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/toktok/c-toxcore/473)
<!-- Reviewable:end -->

iphydf commented:
This breaks windows: https://travis-ci.org/TokTok/c-toxcore/jobs/201035192#L1512

Diadlo commented:
Try to remove all this includes and add them into network.h with freebsd ifdef

yurivict commented:
I moved them into natwork.h
But it looks like ifdef __FreeBSD__ should also be dropped, because linux should need the same includes. They probably get in accidentally on linux through some other path.

Diadlo commented:
@yurivict Do we really need to include it directly on linux? If you think so, feel free to replace on UNIX ifdef (but not drop)

yurivict commented:
I updated the patch.
* Moved all network-related includes together into network.h
* Updated the comment there to say "UNIX includes"

On all UNIXes basic network includes should be the same. It's better to keep them in one place - network.h. Before on Linux some of them were included through some other includes despite manpages saying they should be included.

Diadlo commented:
@yurivict In real, this includes should be in `network.cpp`. But `network.h` should provide platform independent interface ([work in progress](https://github.com/TokTok/c-toxcore/pull/461))
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/473#-:-Kd4l8r0KDrVzP5mSQZv:b-y30hzj)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Can confirm, toxcore doesn't build on FreeBSD 11.0 and this PR fixes the issue.

FreeBSD doesn't have `IPV6_ADD_MEMBERSHIP` and `IPV6_DROP_MEMBERSHIP` defined, `IPV6_JOIN_GROUP` and `IPV6_LEAVE_GROUP` should be used instead. There is already a code block in `network.c` that handles this:

```C
#ifndef IPV6_ADD_MEMBERSHIP
#ifdef  IPV6_JOIN_GROUP
#define IPV6_ADD_MEMBERSHIP IPV6_JOIN_GROUP
#define IPV6_DROP_MEMBERSHIP IPV6_LEAVE_GROUP
#endif
#endif
```

The issue is that `netinet/in.h` needs to be included ***before*** this code block, as that's where the `IPV6_JOIN_GROUP` is defined.

You also need to include `sys/socket.h` in `network.h` for `AF_INET6` and `AF_INET` to be defined, which are used in `DHT.c`.

nurupo commented:
@Diadlo do you think we can merge this PR now or we still need to wait for the network refactor? It looks like there is still work to be done in network refactoring since this is an issue

>You also need to include `sys/socket.h` in `network.h` for `AF_INET6` and `AF_INET` to be defined, which are used in `DHT.c`. 

Ideally `DHT.c` would use platform independent network interface, rather than using `AF_INET6` and `AF_INET`.

Diadlo commented:
@nurupo I'm currently working on replacing AF_INET on platform independent TOX_AF_INET. It will be in the next network refactoring PR. So you can merge this PR

nurupo commented:
Then you'd need to refactoring this merge in that next refactoring PR.

Diadlo commented:
Yes, but currently we haven't base for the writing right platform-independent code

iphydf commented:
@yurivict can you rebase on master, and also enable the checkbox that lets collaborators push to the PR branch.

yurivict commented:
I rebased and enabled the checkbox.

thierry-FreeBSD commented:
In 0.1.7 netinet/in.h is still missing in toxcore/network.c

```
--- toxcore/network.c.orig      2017-03-26 13:42:48 UTC
+++ toxcore/network.c
@@ -43,6 +43,10 @@
 #include <mach/mach.h>
 #endif
 
+#ifdef __FreeBSD__
+#include <netinet/in.h>
+#endif
+
 #ifndef IPV6_ADD_MEMBERSHIP
 #ifdef  IPV6_JOIN_GROUP
 #define IPV6_ADD_MEMBERSHIP IPV6_JOIN_GROUP

```

yurivict commented:
```#ifdef __FreeBSD__``` shouldn't be needed. It doesn't complain on linux only because of accidental inclusion somewhere else.

thierry-FreeBSD commented:
v0.1.11 is still unbuildable on FreeBSD:

```
FAILED: CMakeFiles/toxcore-sut.dir/testing/hstox/driver.c.o 
/usr/local/libexec/ccache/cc  -I/usr/local/include -I/usr/local/include/opus -I/usr/local/include/opencv -O2 -pipe  -fstack-protector -fno-strict-aliasing -std=c99 -pedantic -Wall -Wextra -Weverything -Wno-cast-align -Wno-conversion -Wno-covered-switch-default -Wno-disabled-macro-expansion -Wno-documentation-deprecated-sync -Wno-format-nonliteral -Wno-missing-field-initializers -Wno-missing-prototypes -Wno-padded -Wno-parentheses -Wno-return-type -Wno-sign-compare -Wno-sign-conversion -Wno-tautological-constant-out-of-range-compare -Wno-thread-safety-analysis -Wno-type-limits -Wno-undef -Wno-unreachable-code -Wno-unused-macros -Wno-unused-parameter -Wno-vla -Wno-assign-enum -Wno-bad-function-cast -Wno-double-promotion -Wno-gnu-zero-variadic-macro-arguments -Wno-packed -Wno-reserved-id-macro -Wno-shadow -Wno-shorten-64-to-32 -Wno-unreachable-code-return -Wno-used-but-marked-unused -isystem /usr/local/include -isystem /usr/local/include/opus -isystem /usr/local/include -isystem /usr/local/include -isystem /usr/local/include -isystem /usr/local/include -isystem /usr/local/include -isystem /usr/local/include -isystem /usr/local/include -O2 -pipe  -fstack-protector -fno-strict-aliasing -MD -MT CMakeFiles/toxcore-sut.dir/testing/hstox/driver.c.o -MF CMakeFiles/toxcore-sut.dir/testing/hstox/driver.c.o.d -o CMakeFiles/toxcore-sut.dir/testing/hstox/driver.c.o   -c testing/hstox/driver.c
testing/hstox/driver.c:71:49: error: use of undeclared identifier 'S_IRUSR'
    int fd = open(filename, O_WRONLY | O_CREAT, S_IRUSR | S_IWUSR);
                                                ^
testing/hstox/driver.c:71:59: error: use of undeclared identifier 'S_IWUSR'
    int fd = open(filename, O_WRONLY | O_CREAT, S_IRUSR | S_IWUSR);
                                                          ^
testing/hstox/driver.c:226:24: error: variable has incomplete type 'struct sockaddr_in'
    struct sockaddr_in servaddr;
                       ^
testing/hstox/driver.c:226:12: note: forward declaration of 'struct sockaddr_in'
    struct sockaddr_in servaddr;
           ^
testing/hstox/driver.c:228:38: error: use of undeclared identifier 'INADDR_ANY'
    servaddr.sin_addr.s_addr = htons(INADDR_ANY);
                                     ^
testing/hstox/driver.c:228:38: error: use of undeclared identifier 'INADDR_ANY'
testing/hstox/driver.c:228:38: error: use of undeclared identifier 'INADDR_ANY'
testing/hstox/driver.c:228:38: error: use of undeclared identifier 'INADDR_ANY'
7 errors generated.

```

nurupo commented:
Why are you trying to build the incomplete Haskell testing framework `testing/hstox`?

nurupo commented:
v0.1.11 builds fine on FreeBSD, we have a Travis CI check for that. It doesn't build everything under the sun though, just the actual library and autotests for it, not `testing/hstox` and others.

thierry-FreeBSD commented:
You're right: removing testing/hstox does it.

nurupo commented:
Oh, I'm sorry, I thought `hstox` wasn't enabled by default and that you were explicitly enabling it by some compile flag, but [apparently](https://github.com/TokTok/c-toxcore/blob/fc0cc08b2115d037551369f69519e95e841643aa/CMakeLists.txt#L352-L367) `hstox` gets build by default if `MSGPACK` was found.

thierry-FreeBSD commented:
Yes, I patched CMakeLists.txt to remove it.

nurupo commented:
Yeah, we should probably make it build on FreeBSD given how it's enabled by default, or disable it by default and hide under a build flag. @robinlinden @iphydf

endoffile78 commented:
@thierry-FreeBSD hstox should compile on FreeBSD when #648 is merged.
